### PR TITLE
[Linux] Fix opening files in external apps

### LINKED ISSUE #1490: "Open Instance folder" button does not work
### Operating System

Linux

### Version of Prism Launcher

7.2

### Version of Qt

6.5.2

### Description of bug

"Open Instance folder" button does not work on Arch with Hyprland, pcmanfm. In other apps like Telegram, Firefox buttons with similar functionality works as expected. 
This does not solved issue https://wiki.archlinux.org/title/File_manager_functionality#Directories_are_not_opened_in_the_file_manager

### Steps to reproduce

press "Open Instance folder" button

### Suspected cause

_No response_

### This issue is unique

- [X] I have searched the issue tracker and did not find an issue describing my bug.

### PR DESCRIPTION
**Pull request by Scrumplex**
Closes https://github.com/PrismLauncher/PrismLauncher/issues/1490

How to test
1. Open a folder path
2. Ensure file manager opens
3. Open a file (edit a component in an instance)
4. Ensure that a text editor opens

### CODE REVIEW

File: launcher/DesktopServices.cpp [hunk 1]
```diff
@@ -96,81 +97,30 @@ bool IndirectOpen(T callable, qint64* pid_forked = nullptr)
...
-    if (!isSandbox()) {
-        return IndirectOpen(f);
-    }
-#endif
-    return f();
+    return openUrl(QUrl::fromLocalFile(QFileInfo(path).absolutePath()));
 }
 
-bool openFile(const QString& path)
+bool openPath(const QString& path, [[maybe_unused]] bool ensureExists)
```
Comment (TheKodeToad) [inline launcher/DesktopServices.cpp]: why maybe_unused?

Comment (Scrumplex) [reply launcher/DesktopServices.cpp]: I just stole that from `openDirectory`

**Review by Trial97 — APPROVED** [review]
Please consider removing `IndirectOpen` and `isSandbox` as they are not used anywhere with your current changes.
The code looks good. This still needs testing as I was never able to reproduce the actual issue.

**Comment by shellheim** [comment]
Currently this is what happens: (Fedora 39, flatpak): 

[Screencast_20231217_140514.webm](https://github.com/PrismLauncher/PrismLauncher/assets/116480892/24ee3e2c-a911-4759-aa16-73822af4d664)

This is what it logs when opening the `launcher root` directory and `instances` directory : 

```

    12.355 D | Opening path QFileInfo(instances)
    12.356 D | Opening URL "file:///home/shell/.var/app/org.prismlauncher.PrismLauncher/data/PrismLauncher"
    19.155 D | Opening path QFileInfo(.)
    19.155 D | Opening URL "file:///home/shell/.var/app/org.prismlauncher.PrismLauncher/data"
```

**Comment by TheKodeToad** [comment]
Why the removal of IndirectOpen? What reason did it exist for - did it fix anything - and will removing it make more bugs present? Or was there something which stopped it working properly which can be changed?

**Comment by Scrumplex** [comment]
> Why the removal of IndirectOpen? What reason did it exist for - did it fix anything - and will removing it make more bugs present? Or was there something which stopped it working properly which can be changed?

I am not quite sure myself. But I wouldn't be surprised if MultiMC's issues stemmed from its ancient Qt version.

**Comment by develcooking** [comment]
Just in advance:
I'm not involved in the Prism Launcher, Multi MC or any of that. I also didn't read most of the code. I just want to open my folder from Prism. So I'm trying by searching and gifting possible answers for @TheKodeToad 's questions

> What reason did it exist for - did it fix anything?

- After searching through the commit history of DesktopServices.cpp I found this out

- This is the first commit that added the DesktopService.cpp file [wrap QDesktopServices and QProcess::startDetached](https://github.com/MultiMC/Launcher/commit/4d0caf6254fdb18f4626a3c7937e64422b40d40c) which I think fixed [LD_LIBRARY_PATH set for MultiMC leaks to other processes](https://github.com/MultiMC/Launcher/issues/1389) the [Qt-9328](https://bugreports.qt.io/browse/QTBUG-9328?focusedCommentId=499889) bug

- Other commits in the file:

- [there](https://github.com/MultiMC/Launcher/commit/bbb3b3e6f6e3c0f95873f22e6d0a4aaf350f49d9#diff-2d51209612ff229bd98f4d1036db3172d4b973c64c3cbe32e2cfdc8a7897655d) also a lot changed
all other 

- [here](https://github.com/MultiMC/Launcher/commit/7179e75e70f4157326d6d97f11a947df576d02e6#diff-829be1184c71e91333bbe3440339319115e35ae9e845ef4ec77793b18b835636) was an update to support FreeBSD 

- all other commits just moved the file around but didn't changed it

- The commit history of the DesktopService.cpp file : [link to commit history of Multi MC](https://github.com/MultiMC/Launcher/commits/7450faaf1355af566515b9a156c704c6cd9a21ff/launcher/DesktopServices.cpp)

> did it fix anything?

**Yes think so. It fixed the self reported bug at issue 1389** LD_LIBRARY_PATH set for MultiMC leaks to other processes](https://github.com/MultiMC/Launcher/issues/1389)

>  and will removing it make more bugs present? Or was there something which stopped it working properly which can be changed?

**I didn't find any while testing it** see  last uploaded video at issue #1490

On my end @Scrumplex [branch](https://github.com/Scrumplex/PrismLauncher.git#branch=fix/open-paths-immediately) works 


I hope this could help at leased a little bit

**Comment by TheKodeToad** [comment]
Sorry! I probably could have figured this out if I put in more effort. I assumed Scrumplex knew though.

File: launcher/DesktopServices.h [hunk 2]
```diff
@@ -3,31 +3,30 @@
...
  * Use this instead of QDesktopServices!
  */
 namespace DesktopServices {
 /**
- * Open a file in whatever application is applicable
+ * Open a path in whatever application is applicable.
+ * @param ensurePathExists Make sure the path exists
  */
-bool openFile(const QString& path);
+bool openPath(const QFileInfo& path, bool ensurePathExists = false);
```
Comment (TheKodeToad) [inline launcher/DesktopServices.h]:
```suggestion
bool openPath(const QFileInfo& path, bool makeFolder = false);
```
maybe rename the parameter so it's clear you shouldn't be using this when opening files

**Review by TheKodeToad — CHANGES_REQUESTED** [review]
I am pretty sure this will reintroduce the bugs which the IndirectOpen impl was meant to fix :/

**Comment by Scrumplex** [comment]
> I am pretty sure this will reintroduce the bugs which the IndirectOpen impl was meant to fix :/

But no other Qt app I know uses IndirectOpen or something similar. This really seems like a legacy workaround that is not relevant nowadays

**Comment by TheKodeToad** [comment]
I will approve if you fix the thing I mentioned - or you do not see it to be necessary!

**Comment by Scrumplex** [comment]
> I will approve if you fix the thing I mentioned - or you do not see it to be necessary!

I was working on it ^^

**Comment by shellheim** [comment]
> I will approve if you fix the thing I mentioned - or you do not see it to be necessary!

It still doesn't work in flatpak though :(

**Comment by Scrumplex** [comment]
> It still doesn't work in flatpak though :(

Do you have any other Qt (5) apps in Flatpak to compare?

**Comment by shellheim** [comment]
There are many qt5 apps which can open files but I don't have any that needs to open folders.

File: launcher/ui/pages/instance/ScreenshotsPage.cpp [hunk 3]
```diff
@@ -325,7 +325,7 @@ void ScreenshotsPage::onItemActivated(QModelIndex index)
         return;
     auto info = m_model->fileInfo(index);
     QString fileName = info.absoluteFilePath();
-    DesktopServices::openFile(info.absoluteFilePath());
+    DesktopServices::openPath(info.absoluteFilePath());
```
Comment (TheKodeToad) [inline launcher/ui/pages/instance/ScreenshotsPage.cpp]:
```suggestion
    DesktopServices::openPath(info);
```
also, delete the fileName variable as it's completely unused.

Comment (Scrumplex) [reply launcher/ui/pages/instance/ScreenshotsPage.cpp]: I love bad code!!!

**Review by TheKodeToad — APPROVED** [review]

**Comment by Trial97** [comment]
Maybe related to: https://github.com/PrismLauncher/PrismLauncher/issues/993

**Pull request merged**